src/NUnitFramework/framework/Constraints/Comparers/StreamsComparer.cs

```diff
@@ -100,6 +110,11 @@ public static EqualMethodResult Equal(object x, object y, ref Tolerance toleranc
                 {
                     yStream.Position = actualPosition;
                 }
+
+#if !NETFRAMEWORK
+                arrayPool.Return(bufferExpected);
+                arrayPool.Return(bufferActual);
```

Comment:
@mithileshz 
Based on the info in the test failures, I am wondering if data from the pooled arrays is persisting across calls. Something you can try is to pass `true` as a second argument here when returning the arrays to the pool to force the content to be cleared between usages :)

Reply:
Ok I think I have finally understood what happens.

Basically there is no gurantee that `ArrayPool.Shared.Rent(size)` gives you an array with values `0`. It is not a replacement for `new byte[size]` which means that when we use the binaryReader to read into these arrays (line 68ish), it will override the first _n_ items in the byte array but not the rest. This causes the `SequenceEqual` method to return false as it will read the whole byte array and compare the two. This causes the unit test to fail as the data is not the same.

Reply:
Good sleuthing on the cause. I'm hesitant to suggest we move to a smaller buffer size here as 4096 bytes also aligns with many of the internal .NET and Windows buffer sizes, which carries it's own performance benefits.

I think `ArrayPool` could still be a viable option, we may just need to manually create spans over only the consumed part of the buffer and then pass that into `MemoryExtensions.SequenceEqual` instead. Something like the below works on my machine.

```diff
+if (MemoryExtensions.SequenceEqual<byte>(bufferExpected, bufferActual))
-if (MemoryExtensions.SequenceEqual(bufferExpected.AsSpan(0, readExpected), bufferActual.AsSpan(0, readActual)))
```

Reply:
Ah yes I didn't think about this solution!
It works with the above.
I've updated the PR and here's the benchmark:

```
| Method       | Size  | Mean     | Error     | StdDev    | Ratio | RatioSD | Gen0   | Gen1   | Allocated | Alloc Ratio |
|------------- |------ |---------:|----------:|----------:|------:|--------:|-------:|-------:|----------:|------------:|
| NewByteArray | 32768 | 1.829 μs | 0.0315 μs | 0.0768 μs |  1.00 |    0.00 | 0.5112 | 0.0114 |    8560 B |        1.00 |
| StackAlloc   | 32768 | 1.370 μs | 0.0266 μs | 0.0249 μs |  0.76 |    0.03 | 0.0191 | 0.0019 |     320 B |        0.04 |
| ArrayPool    | 32768 | 1.503 μs | 0.0203 μs | 0.0190 μs |  0.83 |    0.02 | 0.0191 | 0.0019 |     320 B |        0.04 |
```